Maximal angles between convex cones

orlitzky opened the pull request:
My colleagues are writing about this stuff again. Let's migrate https://github.com/sagemath/sage/issues/29169 from the old Trac branch.

Relative to _develop_:
* Add a new reference
* Add a new module `src.sage.geometry.cone_critical_angles` for the implementation
* Add a `max_angle()` method to the class in `src.sage.geometry.cone` for the user interface

Relative to the old trac branch:

* Update the algorithm to take https://www.heldermann.de/JCA/JCA31/JCA311/jca31015.htm into consideration
* Add that paper as a reference
* Drop the more-general critical angles computation. It isn't nearly as nice as the maximal angle method.

TODO:

* ~~Revert commit 95e57636c9 after checking the docs~~

orlitzky committed:
Trac #29169: add the Or2020 reference.

orlitzky committed:
Trac #29169: add the geometry/cone_critical_angles module.

orlitzky committed:
Trac #29169: add the critical_angles and max_angle methods.

orlitzky committed:
Trac #29169: tweak two max/critical angle error messages.

The phrase "this cone cannot be..." can be misinterpreted as making a
value judgment about the given cone, rather than a blanket statement
about the argument (self) to these two functions. This commit changes
it to "cone should not be..." to indicate that the user should not
pass in either a trivial cone or the ambient space.

orlitzky committed:
Trac #29169: improve error message for _random_admissible_cone().

The error message output when _random_admissible_cone() receives an
invalid ambient dimension was a full sentence. This commit pares it
down to just the important part; amely, that there are no nontrivial
cones in the ambient dimension.

orlitzky committed:
Trac #29169: remove two unused example lines.

The line "G_index_sets = list(gevp_licis(G))" appears twice in EXAMPLES,
but the variable "G_index_sets" is not used thereafter. It was most
likely copy/pasted into the example by mistake. This commit removes it.

orlitzky committed:
Trac #29169: take advantage of improvements in trac #30605.

Thanks to some recent improvements in cone containment testing, we no
longer have to reimplement it for vectors with irrational entries
ourselves.

orlitzky committed:
Trac #29169: remove set_random_seed() calls.

The sage library is now doctested with a random random seed (imagine),
making most calls to set_random_seed() superfluous. We remove a few
from the cone critical angle code here.

orlitzky committed:
Trac #29169: fix a few typos.

orlitzky committed:
Trac #29169: convert solve_gevp* functions to generators.

The solve_gevp_zero() and solve_gevp_nonzero() functions for cone
critical angles are "internal" and can return generators. Likewise for
the testing function _solve_gevp_naive(). This should be a tiny bit
more efficient.

orlitzky committed:
Trac #29169: add a warning about inexact critical angle arithmetic.

Passing exact=False to either the max_angle() or critical_angles()
method of cones is a bit "dangerous" because we can miss eigenspaces
of dimension >= 2 due to numerical issues. Add a warning to the
docstring about it.

orlitzky committed:
Trac #29169: add an exact=False example for cone max_angle().

orlitzky committed:
Trac #29169: clean up two import lists in cone_critical_angles.py.

Two tests in this module imported the gevp_licis() function but then
did not use it. No longer.

orlitzky committed:
Trac #29169: use matroid implementation for gevp_licis().

We already have a matroid method that does more or less what the new
gevp_licis() function does in cone_critical_angles.py. It's faster to
use the matroid implementation and then tweak its output slightly, so
let's do that instead.

orlitzky committed:
Trac #29169: switch from QQbar to AA for cone critical angles.

For reasons I've long forgotten, the cone max/critical angles
algorithms were using QQbar for "rationals with roots." Switching it
to AA causes no problems now, and is clearly preferable (the only
eigenvalues we compute are of positive-definite matrices).

orlitzky committed:
Trac #29169: use base ring's zero for the zero-eigenvalue subproblem.

This should make the field of "zero" consistent with all the non-zero
eigenvalues.

orlitzky committed:
Trac #29169: return immutable vectors from eigenvalue problems.

This lets the user (and us, in the tests) stick the results into a set
or any other structure whose elements must be hashable.

orlitzky committed:
Trac #29169: replace _lists_equivalent() with set equality.

Now that all eigenvalue problems should be solved over the same field,
and now that we're returning immutable eigenvectors from those
problems, using set equality (as oposed to element-wise ==) becomes
possible.

orlitzky committed:
Trac #29169: kill periods at the end of INPUT/OUTPUT list items.

orlitzky committed:
Trac #29169: kill spaces around [ list comprehensions ].

orlitzky committed:
Trac #29169: change "irrelevant" to "ignored" in a docstring.

orlitzky committed:
src/doc/en/reference/references/index.rst: add the Or2024 reference

orlitzky committed:
src/sage/geometry/cone.py: consolidate "long time"

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: consolidate "long time"

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: max angle continuity fallback

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: fix critical_angles comment

orlitzky committed:
src/sage/geometry/cone*.py: drop critical_angles() method

Recent research has shown that (in many cases) the maximal angle does
not change too much if the cones are perturbed. This creates even more
of a safety net when computing the maximal angle, because falling back
to inexact arithmetic "perturbs" the cones.

The critical angle function, however, is not lower-semicontinuous as a
set-valued map. I don't like having a "debug" parameter that tells you
you're getting wrong results, and a user interface that requires you
to understand the subtle differences between upper- and
lower-semicontinuity for set-valued maps is not awesome. I think it's
best if we leave this out for now.

orlitzky committed:
src/sage/geometry/cone.py: document the inexact fallback for max_angle()

The max_angle() can fall back to exact=False if exact=True runs in to
(known) issues. The [Or2024] reference discusses when/why this is OK.

orlitzky committed:
src/sage/geometry/cone*.py: fix pycodestyle warnings

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: don't import from "all"

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: drop unused _is_critical()

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: rewrap long lines

orlitzky committed:
src/sage/geometry/cone.py: rewrap long lines

orlitzky committed:
src/sage/geometry/cone.py: allow the full space in max_angle()

This special case is handled by the P.intersection(-Q) check before we
have to think too hard about whether or not it would hurt anything.

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: the full space is admissible

Allow _random_admissible_cone() to generate the entire ambient
space. This only affects a few test cases where it doesn't matter.

orlitzky committed:
src/sage/geometry/cone.py: update a max_angle() test

Maxima no longer crashes, so the description was wrong, and we can
avoid numerical comparisons with simplify() and a round trip through
sympy.

orlitzky committed:
src/sage/geometry/cone.py: test the full space in max_angle()

orlitzky committed:
src/sage/geometry/cone*.py: max_angle() documentation improvements

orlitzky committed:
src/sage/geometry/cone.py: one max_angle() test case is still crashy

Witness:

  sage: n = 4 ## line 6313 ##
  sage: K = cones.schur(n) ## line 6314 ##
  sage: bool( K.max_angle()[0].simplify() == ((n-1)/n)*pi ) ## line 6315 ##
  ------------------------------------------------------------------------
  /usr/lib/python3.11/site-packages/cysignals/signals.cpython-311-x86_64-
  linux-gnu.so(+0x8950)[0x7fa8349c1950]

So let's go through sympy in the n=4 case, too.

orlitzky committed:
src/sage/geometry/cone.py: bring back the max_angle() warning

We should explain why it's not a good idea to _start_ with
exact=False.

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: minor doc rendering issues

orlitzky committed:
src/sage/geometry/cone.py: both -> each, in a docstring

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: delete "." in the module title

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+
+    INPUT:
+
+    - ``M`` -- the matrix whose `(i,j)^{th}` entry is the inner
```
```suggestion
    - ``M`` -- the matrix whose `(i,j)`-th entry is the inner
```

orlitzky replied:
Done, but are you sure this works? I started with `` `(i,j)`th `` and later changed it because it was being mis-parsed.

kwankyu replied:
The hyphen does the trick.

orlitzky replied:
Ok, thanks for the confirmation.

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+    ``G`` that correspond to linearly-independent sets (of columns of
+    ``G``).
+
+    Mnemonic: linearly-independent column-index subsets (LICIS).
```
```suggestion
    ``G`` that correspond to linearly independent sets (of columns of
    ``G``).

    Mnemonic: linearly independent column-index subsets (LICIS).
```
(throughout)

orlitzky replied:
Done.

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+    A = matrix.block([ [ ZZ(0),              M[I,J] ],
+                       [ M.transpose()[J,I], ZZ(0)  ] ])
+    B = matrix.block([ [ GG[I,I],    ZZ(0) ],
+                       [ ZZ(0), HH[J,J]  ] ])
```
align blocks?

orlitzky replied:
Done.

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+        sage: from itertools import chain
+        sage: from sage.geometry.cone_critical_angles import (
+        ....:   _normalize_gevp_solution,
+        ....:    _random_admissible_cone,
```
```suggestion
        ....:   _random_admissible_cone,
```

orlitzky replied:
Done.

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+    Compute the matrix `M` whose `(i,j)^{th}` entry is the
+    inner product of ``gs[i]`` and ``hs[j]``.
+
+    This is the "generalized gram matrix" appearing in Proposition 6
```
```suggestion
    This is the "generalized Gram matrix" appearing in Proposition 6
```

orlitzky replied:
Done.

mkoeppe commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,978 @@
+        sage: M = G.transpose()*H
+        sage: G_index_sets = list(gevp_licis(G))
+        sage: H_index_sets = list(gevp_licis(H))
+        sage: all( set( _normalize_gevp_solution(s) for s in
```
Beware the style police

orlitzky replied:
I'm not sure what you meant so I've reindented these two examples and have probably made them worse.

orlitzky replied:
I may have accidentally made this acceptable to PEP8. In any case, the style police haven't filed any charges.

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: no "-" in linearly independent

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: align matrix blocks

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: consistent doctest indentation

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: capitalize Gram

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: reindent two doctests

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: avoid superscript "th" in prose

tscrim commented on src/sage/geometry/cone.py:
```diff
@@ -6194,6 +6194,205 @@ def Z_operators_gens(self):
+        If it fails when ``exact`` is ``True`` and if each of your
+        cones :meth:`is_strictly_convex`, then a second attempt will
+        be made using inexact arithmetic. (This sometimes avoids the
+        problem [Or2024]_). If the computation fails when your cones
```
```suggestion
        If it fails when ``exact`` is ``True`` and if each of the
        cones :meth:`is_strictly_convex`, then a second attempt will
        be made using inexact arithmetic. (This sometimes avoids the
        problem noted in [Or2024]_). If the computation fails when the cones
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone.py:
```diff
@@ -6194,6 +6194,205 @@ def Z_operators_gens(self):
+          this computation is only known to be stable when both of
+          your cones are strictly convex (or when one of them is the
+          entire space, but the maximal angle is obviously `\pi` in
+          that case).
```
```suggestion
            Using inexact arithmetic (``exact=False``) is faster, but
            this computation is only known to be stable when both of
            your cones are strictly convex (or when one of them is the
            entire space, but the maximal angle is obviously `\pi` in
            that case).
```

orlitzky replied:
Done, and changed "your cones" to "the cones" here too.

tscrim commented on src/sage/geometry/cone.py:
```diff
@@ -6194,6 +6194,205 @@ def Z_operators_gens(self):
+        OUTPUT:
+
+        A triple `\left( \theta_{\text{max}}, u, v \right)`
+        containing,
```
```suggestion
        A triple `\left( \theta_{\max}, u, v \right)`
        containing:
```

orlitzky replied:
When the subscript is a word and not a variable I prefer to see it in text face, but I've no idea if I'm being (in)consistent and don't want to waste time checking, so I've gone ahead and changed this.

tscrim replied:
Note that I had `\max`, not `max`. So it will display as text. I agree with you that text formatting is better.

orlitzky replied:
Ah, I missed the `\` in the diff. Thanks, I'll put it back.

tscrim commented on src/sage/geometry/cone.py:
```diff
@@ -6194,6 +6194,205 @@ def Z_operators_gens(self):
+        Q = other
+
+        from sage.geometry.cone_critical_angles import max_angle
+        return max_angle(P, Q, exact, epsilon)
```
```suggestion
        from sage.geometry.cone_critical_angles import max_angle
        return max_angle(self, other, exact, epsilon)
```
I don't think that helps. Since the main algorithm is in the `max_angle` function, that is where matching the paper is useful.

orlitzky replied:
In hindsight I agree. Fixed!

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+extended in [SS2016]_ to two cones as a generalization of the
+principal angle between two vector subspaces.
+
+Seeger and Sossa proposed an algorithm to find maximal angles, and
```
```suggestion
Seeger and Sossa proposed an algorithm in [SS2016]_ to find maximal angles, and
```
Good to be explicit IMO.

orlitzky replied:
No problem, done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+polyhedral convex cones. Even though all of the functions in this
+module are internal, some are more internal than others. There are a
+few functions that are used only in doctests, and not by any code that
+an end-user would run. Breaking somewhat with tradition, only those
+methods have been prefixed with an underscore.
```
I don't understand the last sentence. Although there are benefits to exposing the functions to the public doc, even for technical functions. They still wouldn't be accessible from Sage's API. So I would probably not make them hidden in general (unless perhaps they are just a trivial helper function).

orlitzky replied:
Typically a module will have functions that are part of the public API, and internal (underscore) methods that are not. But in this case, the entire module should be considered internal:

> This module is internal to SageMath

Moreover it contains an extra class of functions beyond just the public and internal ones, namely functions used only for testing. Those testing-only functions are certainly internal, so I've given them an underscore prefix. But now what should we do with the non-testing functions, all of which are *also* internal?

If we give them all underscores, then there's no way to differentiate between the testing-only functions and the ones that are called in real code by way of `max_angle()`. So, in this case, I've left the underscore off the internal functions that are used in real code.

Nothing is truly internal in python anyway so this seemed like the least-bad option to me. I'm open to suggestions though.

tscrim replied:
Okay, thanks. Now I understand.

I would probably lean towards making them all visible so they appear in the documentation. That way someone looking at this module's doc will know they are there and could potentially use them as part of their own testing. It seems reasonable to expose them in the documentation (especially considering how much effort you spent documenting these hidden functions). Anyways, it's okay to leave it the way it is too. I don't have a very strong opinion on this.

orlitzky replied:
They won't appear in the documentation anyway unless we add the module to the doc index (see the reverted TODO in the PR description). I personally like seeing the docs but I don't want to give anyone the impression that these are supported/stable interfaces. Not that I foresee major changes, but I'd like to reserve the right to hack away parts of  `cone_critital_angles.py` without deprecation warnings so long as `max_angle()` remains untouched.

tscrim replied:
I would add it to the doc index. You're right that having it exposed like that would mean, strictly speaking, we should deprecate stuff. Although we can be flexible, but you can also add some redundancy with an additional statement saying everything except `max_angle()` in this file might change without a deprecation.

orlitzky replied:
I envy your optimism :)

I've added it back to the doc index along with a big warning at the top of the module doc stating that everything is subject to change.

I think that's it, thanks for taking a look at this.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    # There's a fast implementation of this for matroids, but we need
+    # to drop the empty set from its output and convert the rest to
+    # lists that are all sorted in the same order.
+    return map(sorted, filter(bool,LinearMatroid(G).independent_sets()))
```
```suggestion
    return map(sorted, filter(bool, LinearMatroid(G).independent_sets()))
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+                       [ M.transpose()[J,I], ZZ(0)  ] ])
+    B = matrix.block([ [ GG[I,I], ZZ(0)   ],
+                       [ ZZ(0),   HH[J,J] ] ])
+    M = B.inverse() * A
```
```suggestion
    A = matrix.block(ZZ, [[ ZZ.zero(), M[I,J] ],
                       [M[I,J].transpose(), ZZ.zero()]])
    Binv = matrix.block_diagonal(ZZ, [GG[I,I].inverse(), HH[J,J].inverse()])
    M = Binv * A
```
These are mostly mirco-optimizations with more direct cache usage. It is better to transpose and invert the smaller matrices too. PEP8 also doesn't like such alignments... (yes, I think there should be exceptions too...)

orlitzky replied:
I've changed `ZZ(0)` to `ZZ.zero()` since that is a good idea in any case. But this code is inside the `_solve_gevp_naive` function, whose only purpose is to test the two non-naive GEVP functions. As the docs say,

> This is used only for testing, to ensure that the smart way of solving the generalized eigenvalue problem ... returns the same answers as the dumb way.

For that reason I have avoided using any mathematical tricks (no matter how obvious) to simplify the problem. Is it OK to leave the block matrices? I can trim some of the spaces to conform to PEP8.

tscrim replied:
It's not really a trick, it's simply realizing it is a block diagonal matrix. Anyways, if you prefer it, it is okay.

orlitzky replied:
Ok, thanks. I've removed the extra spacing and reindented:

```python
    A = matrix.block([
        [ZZ.zero(), M[I,J]],
        [M.transpose()[J,I], ZZ.zero()]
    ])
    B = matrix.block([
        [GG[I,I], ZZ.zero()],
        [ZZ.zero(), HH[J,J]]
    ])
```

PEP8 is okay with that.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+
+def solve_gevp_zero(M, I, J):
+    r"""
+    Solve the generalized eigenvalue problem in Theorem 3 for a zero
```
```suggestion
    Solve the generalized eigenvalue problem in Theorem 3 [Or2020]_ for a zero
```
Otherwise, which Theorem 3?

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+      `G` that appears in Theorem 3
+
+    - ``J`` -- a linearly independent column-index set for the matrix
+      `H` that appears in Theorem 3
```
```suggestion
    - ``I`` -- a linearly independent column-index set for the matrix
      `G` that appears in Theorem 3 [Or2020]_

    - ``J`` -- a linearly independent column-index set for the matrix
      `H` that appears in Theorem 3 [Or2020]_
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    # isn't smart enough to figure out a basis for the product. So,
+    # we use the direct sum and then chop it up.
+    xi_space = (M.transpose()[J,I]).right_kernel()
+    eta_space = (M[I,J]).right_kernel()
```
```suggestion
    MIJ = M[I,J]
    xi_space = MIJ.left_kernel()
    eta_space = MIJ.right_kernel()
```

orlitzky replied:
Done, thanks

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+
+    - ``HH`` -- the matrix whose `(i,j)`-th entry is the inner product
+      of `h_{i}` and `h_{j}`, which are in turn the `i`-th and `j`-th
+      columns of the matrix `H` in Theorem 3
```
```suggestion
      columns of the matrix `G` in Theorem 3 [Or2020]_

    - ``HH`` -- the matrix whose `(i,j)`-th entry is the inner product
      of `h_{i}` and `h_{j}`, which are in turn the `i`-th and `j`-th
      columns of the matrix `H` in Theorem 3 [Or2020]_
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+      `G` that appears in Theorem 3
+
+    - ``J`` -- a linearly independent column-index set for the matrix
+      `H` that appears in Theorem 3
```
```suggestion
      `G` that appears in Theorem 3 [Or2020]_

    - ``J`` -- a linearly independent column-index set for the matrix
      `H` that appears in Theorem 3 [Or2020]_
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    - ``multiplicity`` is the dimension of the eigenspace associated
+      with ``eigenvalue``
+
+    Note that we don't return a basis for each eigenspace along with
```
```suggestion
    Note that we do not return a basis for each eigenspace along with
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+        ....: )
+        True
+
+    According to Proposition 7, the only eigenvalues that arise when
```
```suggestion
    According to Proposition 7 [Or2020]_, the only eigenvalues that arise when
```

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+        # the "eta problem."
+        yield from ( (l, xi, eta, m)
+                     for (l, eta, xi, m)
+                     in solve_gevp_nonzero(HH,GG,M.transpose(),J,I) )
```
```suggestion
        yield from ((l, xi, eta, m)
                     for (l, eta, xi, m)
                     in solve_gevp_nonzero(HH, GG, M.transpose(), J, I))
```
PEP8

orlitzky replied:
Done

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+                     in solve_gevp_nonzero(HH,GG,M.transpose(),J,I) )
+    else:
+        G_I_pinv_H_J = GG[I,I].inverse_positive_definite()*M[I,J]
+        H_J_pinv_G_I = HH[J,J].inverse_positive_definite()*M.transpose()[J,I]
```
```suggestion
        MIJ = M[I,J]
        G_I_pinv_H_J = GG[I,I].inverse_positive_definite() * MIJ
        H_J_pinv_G_I = HH[J,J].inverse_positive_definite() * MIJ.transpose()
```

orlitzky replied:
Done, thanks

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+                    for l in [-sigma.sqrt(), sigma.sqrt()]:
+                        eta = ~l * H_J_pinv_G_I*xi
+                        eta.set_immutable()
+                        yield (l, xi, eta, m)
```
```suggestion
        for (sigma, xis, m) in L.eigenvectors_right():
            if sigma > 0:
                sigma_sqrt = sigma.sqrt()
                inv_sqrt = ~sigma_sqrt
                sqrts = [-sigma_sqrt, sigma_sqrt]
                inv_sqrts = [-inv_sqrt, inv_sqrt]
                for xi in xis:
                    for l, li in zip(sqrts, inv_sqrts):
                        eta = li * H_J_pinv_G_I * xi
                        eta.set_immutable()
                        yield (l, xi, eta, m)
```
Trying to avoid repeating computations.

orlitzky replied:
Do you think it's worth pulling out the `zip` as well? Something like,

```python
pm_sqrt_inv_pairs = [
    (-sigma_sqrt, -inv_sqrt),
    (sigma_sqrt, inv_sqrt),
]
for xi in xis:
    for l, li in pm_sqrt_inv_pairs:
```

I can't say for sure now that `zip` returns an iterator.

orlitzky replied:
Avoiding `zip` looks like it saves about a microsecond, but at that scale it's impossible to say for sure.

orlitzky replied:
Done, thanks

tscrim replied:
Pulling out the zip (as a zip) wouldn't work because of the iterator, but you're right that it should be better to just do it manually since it is a very small list. In either case, the biggest thing is making sure the `if sigma > 0:` is outside the `for` loops.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    min_v = hs[0]
+    min_ip = min_u.inner_product(min_v)
+
+    M = []
```
```suggestion
    M = [None] * len(gs)
```
It is faster to initialize the list to the correct size if known.

orlitzky replied:
It doesn't seem to be faster in this case?

tscrim replied:
Definitely a micro-optimization, and the benefit depends on the size of the list:
```python
sage: def test1(n):
....:     L = [None]*n
....:     for i in range(n):
....:         L[i] = i
sage: def test2(n):
....:     L = []
....:     for i in range(n):
....:         L.append(i)
sage: %timeit test1(5)
516 ns ± 14.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test2(5)
426 ns ± 9.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test1(20)
923 ns ± 25.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test2(20)
969 ns ± 38.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test1(1000)
34.6 µs ± 1.92 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
sage: %timeit test2(1000)
43.9 µs ± 2.18 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

orlitzky replied:
Ah, all my examples were small. In that case I'd rather leave this alone. The algorithm is combinatorial in the size of `gs` and `hs`,  so you're not going be solving problems where `M` is large enough to see any benefit here.

tscrim replied:
Okay, no problem.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+                min_ip = M_i[j]
+                min_u = gs[i]
+                min_v = hs[j]
+        M.append(M_i)
```
```suggestion
        M[i] = M_i
```

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+            if (M_i[j] < min_ip):
+                min_ip = M_i[j]
+                min_u = gs[i]
+                min_v = hs[j]
```
```suggestion
        M_i = [None] * len(hs)
        for v in hs:
            val = M_i.append(gs[i].inner_product(v))
            if val < min_ip:
                min_ip = val
                min_u = gs[i]
                min_v = v
```
Similar such changes.

orlitzky replied:
list append doesn't return a value... or am I missing something? I could still assign `M_i[j]` to a variable to avoid re-indexing but surely by itself that's overkill.

tscrim replied:
Ah, sorry, that was a mistake. It should have been
```python
for j, v in enumerate(hs):
    val = gs[i].inner_product(v)
    M_i[j] = val

orlitzky replied:
Since I'm keeping `M = []` and the append, I was able to omit the indices entirely. Looks like a small improvement, thanks!

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    u = G_I*xi_hat
+    rhs = u - cos_theta*H_J*eta_hat
+    if any( x < -epsilon for x in H_J_c_T*rhs):
+        return infeasible_result
```
```suggestion
    v = H_J * eta_hat
    rhs = v - cos_theta * G_I * xi_hat

    if any(x < -epsilon for x in G_I_c_T * rhs):
        return infeasible_result

    u = G_I * xi_hat
    rhs = u - cos_theta * H_J * eta_hat
    if any(x < -epsilon for x in H_J_c_T * rhs):
        return infeasible_result
```
PEP8

orlitzky replied:
I fixed some of these, but PEP8 actually recommends writing products without spaces when addition/subtraction (or any lower-precedence operation) are involved: https://peps.python.org/pep-0008/#other-recommendations

As a result, `v - cos_theta*G_I*xi_hat` is probably correct -- it's one of the examples they give. The expression,

```python
x < -epsilon for x in H_J_c_T*rhs
```

is more ambiguous, but in theory you could be multiplying the result of the list comprehension by `rhs`. Writing `H_J_c_T*rhs` with no spaces helps indicate (to me, anyway) that it's one quantity and binds stronger than the `for foo in bar`. I'm still happy to change it if you care, though.

mkoeppe replied:
> As a result, `v - cos_theta*G_I*xi_hat` is probably correct

I strongly agree. It's correct per PEP 8 and much more readable than the fully "spaced out" version.

> Writing `H_J_c_T*rhs` with no spaces

This one is trickier to justify per PEP 8, but I still support it; not with the explanation that you give, but rather because these terms (products) appear right next to polynomial expression where the monomials are written in this way without spaces. Keeping the same style for the same terms aids readability.

tscrim replied:
At the risk of continuing to bikeshed, the `H_J_c_T * rhs` needs the spacing by PEP8 as the `*` is the highest priority operator. It's not a consistency thing.

I  have no objections to not having the spaces for the other as it is correct. (I think it is a bit of an overreach to say PEP8 prefers it, much less recommends it.) I probably wasn't paying close enough attention here. (Although I do find it easier to read with all the underscores in the variable names, but its not important.)

orlitzky replied:
I've just added the spaces.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    else:
+        GG = G.transpose()*G
+        H = matrix.column(hs)
+        HH = H.transpose()*H
```
```suggestion
        GG = G.transpose() * G
        H = matrix.column(hs)
        HH = H.transpose() * H
```
PEP8

orlitzky replied:
Fixed

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+    big_eigenspaces = []
+
+    for I in G_index_sets:
+        G_I = G[range(n),I]
```
Wouldn't `G.matrix_from_columns(I)` be faster here? Similar for other places below.

orlitzky replied:
Yep. I didn't know that method was there.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+            H_J = H[range(n),J]
+            H_J_c_T = H[range(n),J_complement].transpose()
+
+            for (cos_theta,xi,eta,mult) in solve_gevp_nonzero(GG,HH,M,I,J):
```
```suggestion
            for (cos_theta,xi,eta,mult) in solve_gevp_nonzero(GG, HH, M, I, J):
```
PEP8

orlitzky replied:
I'm happy to change it either way, but does PEP8 actually suggest putting spaces after the commas? I've been looking and don't see it.

tscrim replied:
I'm surprised this is not in there. I could have sworn it was a Python style coding convention. It is definitely a very widely used principle that commas should generally be followed by a whitespace. This might be a Sage convention...

orlitzky replied:
OK, I made this change in a few places.

tscrim commented on src/sage/geometry/cone_critical_angles.py:
```diff
@@ -0,0 +1,998 @@
+                elif mult > 1:
+                    # Save this for later. The eigenvalue cos_theta
+                    # might be so big that we can ignore it.
+                    big_eigenspaces.append( (cos_theta, xi, eta, mult) )
```
```suggestion
                    big_eigenspaces.append((cos_theta, xi, eta, mult))
```
PEP8

orlitzky replied:
Fixed

tscrim reviewed (changes requested):
Hopefully these changes are all speed improvements. I didn't check carefully that all statements are equivalent, but hopefully the intent is clear. Let me know if you have questions.

orlitzky committed:
src/sage/geometry/cone.py: don't define P,Q in max_angle()

These names are helpful in the cone_critical_angles module, but not in
max_angle() which immediately passes them off to another method.

orlitzky committed:
src/sage/geometry/cone.py: "your cones" -> "the cones" in max_angle() docs

orlitzky committed:
src/sage/geometry/cone.py: add "noted in" before a citation

orlitzky committed:
src/sage/geometry/cone.py: reindent the WARNING block in max_angle()

orlitzky committed:
src/sage/geometry/cone.py: don't put \text around "max" subscript

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: add another SS2016 citation

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: misc PEP8 corrections

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: another PEP8 fix

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: prefer ZZ.zero() to ZZ(0)

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: add more Or2020 citations

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: change one "don't" to "do not"

orlitzky committed:
src/sage/geometry/cone.py: fix {max} -> {\max} in subscripts

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: add spaces to solve_gevp* calls

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: add spacing to a "filter" call

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: no extra spaces in block matrices

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: micro-optimize solve_gevp_nonzero()

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: use matrix_from_columns()

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: optimize M[I,J] indexing

Using the identity M[I,J].transpose() == M.transpose()[J,I], we can
avoid transposing M itself and instead transpose the smaller M[I,J]
when M[I,J] is already known.

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: more spaces around *

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: avoid indexing in a loop

orlitzky committed:
src/doc/en/reference/discrete_geometry/index.rst: add cone critical angles

This is an internal module, but it's fully documented and nice to have
access to the implementation behind the max_angle() function. Despite
being documented, the functions in cone_critical_angles.py are subject
to change at any time. This itself is loudly documented.

orlitzky committed:
src/sage/geometry/cone_critical_angles.py: add another "internal" warning

The existing warning about this module being internal is easy to
miss. This one should appear in bright red at the top of the module
documentation.

tscrim reviewed (approved):
Okay, thank you. Then let it be so.
